src/xform_to_datalog/datalog_facts.cc

```diff
@@ -0,0 +1,149 @@
+      const std::string &particle_spec_name = particle_proto.spec_name();
+      CHECK(!particle_spec_name.empty())
+        << "Particle with empty spec_name field not allowed.";
+      const std::string particle_name =
+          absl::StrCat(particle_spec_name, "#", particle_num++);
+
+      absl::flat_hash_map<ir::AccessPathRoot, ir::AccessPathRoot>
+          instantiation_map;
+      for (const arcs::HandleConnectionProto &connection_proto :
+        particle_proto.connections()) {
+        std::string handle_spec_name = connection_proto.name();
+        CHECK(!handle_spec_name.empty())
+          << "Handle connection with empty name field not allowed.";
+        std::string handle_name = connection_proto.handle();
+        CHECK(!handle_name.empty())
+          << "Handle with empty handle field not allowed.";
+
+        raksha::ir::HandleConnectionSpecAccessPathRoot spec_handle_root(
+            particle_spec_name, handle_spec_name);
+        raksha::ir::HandleConnectionAccessPathRoot
+          instantiated_handle_root(
+              recipe_name, particle_name, std::move(handle_spec_name));
+        raksha::ir::HandleAccessPathRoot handle_root(
+            recipe_name, std::move(handle_name));
+        CHECK(connection_proto.has_type())
+          << "Handle connection with absent type not allowed.";
+        std::unique_ptr<types::Type> connection_type =
+            types::Type::CreateFromProto(connection_proto.type());
+        ir::AccessPathSelectorsSet access_path_selectors_set =
+            connection_type->GetAccessPathSelectorsSet();
+
+        // For each Selector path from a root to a leaf, draw dataflow edges
+        // to and from the AccessPaths rooted at the handle at at the
+        // particle's handle connection. In the case that the handle
+        // connection is input only, the edge out will be unreachable from
+        // the rest of the graph. In the case that the connection is output
+        // only, the edge in will be unreachable from the rest of the graph.
+        // In the case that the connection is bidirectional, both edges will
+        // be useful. Note that while the additional edges are, in some
+        // sense, "wasted", the worst case is that we draw twice as many
+        // edges as we need to, which doesn't affect asymptotic complexity.
+        // In particular, the dataflow graph remains correct, if not as
+        // minimal as it could be.
+        for (const ir::AccessPathSelectors &selectors :
+             ir::AccessPathSelectorsSet::CreateAbslSet(
+                 access_path_selectors_set)) {
+          ir::AccessPath handle_access_path(
+              ir::AccessPathRoot(handle_root), selectors);
+          ir::AccessPath handle_connection_access_path(
+              ir::AccessPathRoot(instantiated_handle_root), selectors);
+          result_edges.push_back(
+              ir::Edge(handle_access_path, handle_connection_access_path));
+          result_edges.push_back(
```

Comment:
We should add these only if this is a write/readwrite connection?

Reply:
Done.